pkgs/racket-doc/scribblings/reference/treelists.scrbl

```diff
@@ -0,0 +1,758 @@
+results. For a constant-time @racket[proc], this operation takes
+@math{O(N)} time.
+
+@examples[
+#:eval the-eval
+(define items (treelist 1 "a" 'apple))
+(treelist-for-each items println)
+]}
+
+@defproc[(treelist-member? [tl treelist?] [v any/c] [eql? (any/c any/c . -> . any/c) equal?]) boolean?]{
+
+Checks each element of @racket[tl] with @racket[eql?] and @racket[v]
+(with @racket[v] the second argument) until the result is a true
+value, and then returns @racket[#t]. If no such element is found, the
+result is @racket[#false]. For a constant-time @racket[eql?], this
+operation takes @math{O(N)} time.
+
+@examples[
+#:eval the-eval
+(define items (treelist 1 "a" 'apple))
+(treelist-member? items "a")
+(treelist-member? items 1.0 =)
+(eval:error (treelist-member? items 2.0 =))
+]}
+
+@defproc[(treelist-find [tl treelist?] [pred (any/c . -> . any/c)]) any/c]{
+
+Checks each element of @racket[tl] with @racket[pred] until the result
+is a true value, and then returns that element. If no such element is
+found, the result is @racket[#false]. For a constant-time
+@racket[pred], this operation takes @math{O(N)} time.
+
+@examples[
+#:eval the-eval
+(define items (treelist 1 "a" 'apple))
+(treelist-find items string?)
+(treelist-find items symbol?)
+]}
+
+@defproc[(treelist-sort [tl treelist?]
+                        [less-than? (any/c any/c . -> . any/c)]
+                        [#:key key (or/c #f (any/c . -> . any/c)) #f]
+                        [#:cache-keys? cache-keys? boolean? #f])
+         treelist?]{
+
+Like @racket[sort], but operates on a @tech{treelist} to
+produce a sorted treelist. Sorting takes @math{O(N log N)} time.
+
+@examples[
+#:eval the-eval
+(define items (treelist "x" "a" "q"))
+(treelist-sort items string<?)
+]}
+
+@defproc[(in-treelist [tl treelist?]) sequence?]{
+
+Returns a @tech{sequence} equivalent to @racket[tl].
+@speed[in-treelist "treelist"]
+
+@examples[
+#:eval the-eval
+(define items (treelist "x" "a" "q"))
+(for/list ([e (in-treelist items)])
+  (string-append e "!"))
+]}
+
+@deftogether[(
+@defform[(for/treelist (for-clause ...) body-or-break ... body)]
+@defform[(for*/treelist (for-clause ...) body-or-break ... body)]
+)]{
+
+Like @racket[for/list] and @racket[for*/list], but generating
+@tech{treelists}.
+
+@examples[
+#:eval the-eval
+(for/treelist ([i (in-range 10)])
+  i)
+]}
+
+@defproc[(chaperone-treelist [tl treelist?]
+                             [ref-proc (treelist? exact-nonnegative-integer? any/c . -> . any/c)]
+                             [set-proc (treelist? exact-nonnegative-integer? any/c . -> . any/c)]
+                             [insert-proc (treelist? exact-nonnegative-integer? any/c . -> . any/c)]
+                             [append-proc (treelist? treelist? . -> . treelist?)]
+                             [prop impersonator-property?]
+                             [prop-val any/c] ... ...)
+          (and/c treelist? chaperone?)]{
```

Comment:
I have a scenario in mind that I don't think this API would support, though maybe the answer is that this isn't an intended use-case for chaperones.

I have some code that uses a representation like Zuo's [`string-tree?`](https://docs.racket-lang.org/zuo/zuo-base.html#%28def._%28%28lib._zuo-doc%2Ffake-zuo..rkt%29._string-tree~3f%29%29), particularly to get a more efficient alternative to `append` on lists. The trees are used in fields of other data structures. To avoid repeated checks of potentially large trees across boundaries, I wrap validated trees in a struct.

Treelists would be attractive as a replacement for "string trees" if they could still offer an efficient `append` while improving the performance (and simplifying the implementation) of other operations.

Of course, I could keep using a wrapper struct as I do now, but it would be even nicer to let client code manipulate the treelists directly if I could still avoid too much redundant checking.

I could imagine marking validated treelists with, say, `impersonator-prop:validated-string-treelist`. If `set-proc`, `insert-proc`, or `append-proc` would return a treelist that did not pass my check, it would work either to not apply a chaperone wrapper to the result and or to change the property value (in which case I'd skip further checking in the wrappers). In particular, I would not want to raise an exception for disallowed values, because it would be legitimate for client code to take a treelist from the library to use for its own purposes, in which case it should be allowed to `set`/`insert`/`append` other kinds of values freely.

However, this API doesn't seem to allow adjusting properties on the resulting string-trees (or even intercepting them to register in a hash-table on the side or something).

An alternative strategy could be to chaperone treelists with `set`/`insert`/`append` procedures that *do* raise exceptions, but provide a public procedure to unwrap the chaperone&mdash;except I think providing a public procedure to unwrap the chaperone would also require setting an impersonator property on the resulting chaperone to give access to the unchaperoned version.

Either way, there would still be the problem of chaperones changing the algorithmic complexity of `treelist-append`. I would want appending validated treelists to take O(M log n) time. For "mixed" appends, I would want either (a) to inspect contents only of the treelists that were not previously validated or (b) to unconditionally adjust the impersonator property on the result to mark it as validation-unknown, allowing fast appends at the cost of just one revalidation if/when the resulting treelist crossed back into my library code.

If the exception-raising-chaperones+public-unwrap-function strategy worked, I guess I could provide a `validated-string-treelist-append` to use instead, but at that point I'd have a very strange API without the key benefit of avoiding wrapper structs, i.e. allowing the trees to be manipulated with generic functions on treelists.

Reply:
I'm not sure it covers your case, but this was a useful example to refine `chaperone-treelist` in a familiar direction: when `#f` is supplied instead of a procedure for `ref-proc`, then it can avoid changing the asymptotic performance of `chaperone-append` . (Oh, except that I forgot to implement that. Well, I will before merging.)

Yes, using an extra impersonator property to allow unwrapping is the usual way to go. The mutable-treelist implementation uses that trick itself, but it piggy-backs on the table of chaperone procedures that already uses a property.